syntax = "proto3";

package odc;

// The ODC service definition.
service ODC {
    // Initialize
    rpc Initialize (InitializeRequest) returns (GeneralReply) {}
    // Submit agents. Can be called multiple times in order to submit more agents.
    rpc Submit (SubmitRequest) returns (GeneralReply) {}
    // Activate topology.
    rpc Activate (ActivateRequest) returns (GeneralReply) {}
    // Update topology. Can be called multiple times in order to scale up or down the topology.
    rpc Update (UpdateRequest) returns (GeneralReply) {}
    // Configure
    rpc Configure (ConfigureRequest) returns (GeneralReply) {}
    // Set property
    rpc SetProperty (SetPropertyRequest) returns (GeneralReply) {}
    // Start
    rpc Start (StartRequest) returns (GeneralReply) {}
    // Stop
    rpc Stop (StopRequest) returns (GeneralReply) {}
    // Reset
    rpc Reset (ResetRequest) returns (GeneralReply) {}
    // Terminate
    rpc Terminate (TerminateRequest) returns (GeneralReply) {}
    // Shutdown
    rpc Shutdown (ShutdownRequest) returns (GeneralReply) {}
}

// Request status
enum ReplyStatus {
    UNKNOWN = 0;
    SUCCESS = 1;
    ERROR = 2;
}

// Device type
enum DeviceType {
    ALL = 0;
    RECO = 1;
    QC = 2;
}

// General error
message Error {
    string msg = 1;
    int32 code = 2;
}

// General reply to requests
message GeneralReply {
    string msg = 1;
    ReplyStatus status = 2;
    Error error = 3;
    int32 exectime = 4; // Execution time in ms
    uint64 runid = 5;
}

// Initialize request
message InitializeRequest {
    uint64 runid = 1; 
}

// Submit request
message SubmitRequest {
    // TODO: Add request parameters here
}

// Activate request
message ActivateRequest {
    string topology = 1;
}

// Update request
message UpdateRequest {
    string topology = 1;
}

// Shutdown request
message ShutdownRequest {
    // TODO: Add request parameters here
}

// Set property request
message SetPropertyRequest {
    string key = 1;
    string value = 2;
    string path = 3;
}

//
// FairMQ device state change requests
//

// Configure request
message ConfigureRequest {
    DeviceType device = 1;
}

// Start request
message StartRequest {
    DeviceType device = 1;
}

// Stop request
message StopRequest {
    DeviceType device = 1;
}

// Reset request
message ResetRequest {
    DeviceType device = 1;
}

// Terminate request
message TerminateRequest {
    DeviceType device = 1;
}
